### ISSUE TITLE
Function to get diet of predator

### ISSUE THREAD

**Issue by gustavdelius** [sizespectrum/mizer#43]
@astaaudzi has requested a function that returns the diet of a predator. I believe what she wants is a function like the following:
```
#' Get diet of predator resolved by prey species
#'
#' Calculates the rate at which a predator of a particular species and size consumes 
#' biomass of a prey species, where for this purpose we treat the plankton like another
#' prey species. Thus it performs the same integration as \code{getAvailEnergy()} but
#' does not aggregate over prey species, and multiplies by (1-feeding_level) to get the
#' consumed biomass rather than the available biomass.
#'
#' @param params A MizerParams object
#' @param n An array (species x size) with the abundance density of fish
#' @param n_pp A vector with the abundance of plankton
#' 
#' @return An array (predator species  x predator size x (prey species + plankton) )
#' @export
getDiet <- function(params, n, n_pp) {
    ...
}
```

**Comment by astaaudzi** [comment]
Yes, @gustavdelius this is exactly what I have asked. 

I think such a function would very useful for all mizer users. We have tried different diet explorations, and many users were surprised to find out that their species fill up almost entirely on the background spectrum. If the background spectrum has high regeneration rate (r_pp = 10, for example, which is default in mizer), there is not much depletion happening and people effectively end up running a many parallel single species models. So looking at realised diets would be an important step in model development. 
Ideally, there would also be a function to plot the realised diets by predator size for each predator. This should not be that difficult to write in ggplot.

**Comment by gustavdelius** [comment]
I agree with Asta. The plot of the diet also takes a central role in our shiny app for tuning mizer for an ecosystem that we hope to contribute very shortly. The above function I can contribute on the weekend. I also agree that there should be an associated plot function. However it is not entirely clear what the best visualisation would be. We made a line plot that showed one curve for each prey species describing the rate of consumption of that prey against the predator size, for one specified predator species. @astaaudzi made a panel plot with one panel for each predator species showing the proportion of its diet coming from each prey species. So perhaps the plot function should have a `style` option.

**Comment by astaaudzi** [comment]
I guess both options would be very useful and give different information. A prey might have very high mortality from some predator, but still constitute a negligible proportion of predator's diet (who might still be eating mostly background spectrum). By the way, the panel plot was made by Jon Reum using his code to calculate diet proportions by predator x size x prey x size.

**Comment by Kenhasteandersen** [comment]
In that context I would like to share an unpublished figure that Martin Hartvig made. It shows the diet (as a fraction of maximum consumption rate) of species as a function of size. The species are ordered according to asymptotic size. One clearly sees how the larger species have ontogenetic shifts from one species to another, and how these shifts are often associated by bottlenecks in the food availability and thus growth.

Ken



---------------------------------------------------------------------------------------------------------------
Ken H. Andersen, http://ken.haste.dk<http://ken.haste.dk/>, twitter: @69kno
Professor in theoretical marine ecology, head of section, and deputy director of Centre for Ocean Life http://www.oceanlifecentre.dk<http://www.oceanlifecentre.dk/>

New paper: A trait-based approach to ocean ecology<https://www.google.com/url?q=https://academic.oup.com/icesjms/advance-article/doi/10.1093/icesjms/fsy090/5056134&sa=D&sntz=1&usg=AFQjCNFVkw1WnenGUFvOaKHEPxp8I30XKA>. ICES Journal of Marine Science 2, 65–70 (2018)

[cid:90c1d632-fbba-46e8-a76a-ca245f7ae4d6@win.dtu.dk]

On 7 Mar 2019, at 11.21, Asta Audzi <notifications@github.com<mailto:notifications@github.com>> wrote:


I guess both options would be very useful and give different information. A prey might have very high mortality from some predator, but still constitute a negligible proportion of predator's diet (who might still be eating mostly background spectrum). By the way, the panel plot was made by Jon Reum using his code to calculate diet proportions by predator x size x prey x size.

**Comment by astaaudzi** [comment]
@Kenhasteandersen, did you mean to attach a figure?

**Comment by Kenhasteandersen** [comment]
I did attach it, but it might not go through to the mailing list. Here it is: https://www.dropbox.com/s/6cftmzlq64c2s95/consumption.pdf?dl=0

Ken

---------------------------------------------------------------------------------------------------------------
Ken H. Andersen, http://ken.haste.dk<http://ken.haste.dk/>, twitter: @69kno
Professor in theoretical marine ecology, head of section, and deputy director of Centre for Ocean Life http://www.oceanlifecentre.dk<http://www.oceanlifecentre.dk/>

New paper: A trait-based approach to ocean ecology<https://www.google.com/url?q=https://academic.oup.com/icesjms/advance-article/doi/10.1093/icesjms/fsy090/5056134&sa=D&sntz=1&usg=AFQjCNFVkw1WnenGUFvOaKHEPxp8I30XKA>. ICES Journal of Marine Science 2, 65–70 (2018)

[cid:image001.jpg@01D09493.A377DFB0]

On 7 Mar 2019, at 11.28, Asta Audzi <notifications@github.com<mailto:notifications@github.com>> wrote:


@Kenhasteandersen<https://github.com/Kenhasteandersen>, did you mean to attach a figure?

**Comment by astaaudzi** [comment]
Thanks. This looks good and makes a lot of sense. I think the issues may arise in a multi-species setup, where species w_inf are nowhere near equally spaced, the interaction matrix has a lot of zeroes (because there is no evidence of those species in diets) and PPMRs are very different across species (some planktivores, some predators, etc). 
This is probably quite far from the original idea of size based model in a pelagic system (if I understand it correctly), but it is a direction mizer is now increasingly often used I think. 
In summary, a diet output and plot function would be very useful :)

**Comment by gustavdelius** [comment]
I have added the getDiet() function. I have not yet worked on the plotDiet() function.

**Comment by astaaudzi** [comment]
@gustavdelius , I have just started looking into using this function for my model. As I understand the getDiet will give me the biomass of prey consumed by all predator individuals in a size class? Since this function gives a rate I suppose the units are in g/year?

Just to make sure I understand this properly: 
So to get the amount of prey consumed by a single fish we would need to divide getDiet output by the numbers of predators at a given time step. Otherwise the consumption in grams is decreasing with the size of a predator. Then to get diet proportions we further should divide this by the total amount of prey consumed by each size class of a predator? I imagine it is the last value (diet proportion) that is of most interest to users

**Comment by gustavdelius** [comment]
@astaaudzi , the function currently returns the biomass of prey that a single predator consumes. The documentation of the current version is at https://sizespectrum.github.io/mizer/dev/reference/getDiet.html. Let me know if it should be clarified.

Given what you say about people being interested in the proportion of the diet coming from each prey species rather than the biomass, I propose that we add a boolean parameter `proportion` which causes the function to return the proportion when set to `TRUE`  and the biomass when set to `FALSE`. Does that sound like a good idea?
